### PR TITLE
Feature/core plugins

### LINKED ISSUE #96: Core input plugins
Content types overrides any custom input plugin directories if a Punchcard administrator sets up the cms as a required module. A change needs to happen at the Punchcard level to allow for multiple directories to be sent into content types to search for input plugins.

```gherkin
Feature: Plugin Locations
  As an administrator
  I want a way to access Core plugins
  So that I can use my own plugins, but still have access to a containing application's core plugins.

  Scenario: My input plugins
    Given that I am creating a form using content-types
      And that content-types requires a content-type configuration to create forms
      And that I have some input plugins locally in my application that I need to use in my forms
     When creating my form
     Then I need to use input plugins which are not available via a package manager
      And I need to use input plugins which are inside my application.

  Scenario: Core input plugins
    Given that I am creating a content-type using Punchcard CMS
      And that Punchcard CMS uses content-types to create forms
      And that Punchcard has core input plugins
      And that I need to use a Punchcard input plugin, references, in my form
     When creating a content type configuration for use in Punchcard CMS
     Then I need to use input plugins which are not available via a package manager
      And I need to use input plugins which ship with Punchcard.

  Scenario: Accept multiple directories
    Given that I am creating a form using content-types
     And that multiple input plugin directories are part of my application
      And that I'm using content-types to create forms using plugins local to my application
    When I send a configuration with more than one input plugin directory to content-types
    Then I need all input-plugin directories to have their plugins available to use in my forms.
```

---

**Plugin Locations**
- [ ] My input plugins
- [ ] Core input plugins
- [ ] Accept multiple directories

### PR DESCRIPTION
**Pull request by scottnath**
We need access to plugins that ship with punchcard as well as ones used locally.

---
Resolves #96 

`DCO 1.1 Signed-off-by: Scott Nath <github@scottnath.com>`

### CODE REVIEW

File: lib/utils.js [hunk 1]
```diff
@@ -55,34 +55,68 @@ exports.splitPop = (input, splitter) => {
...
  *
- * @param {object} globalConfig - Configuration object
+ * @param {object} globalConfig - global configuration file contains plugin directories
  *
- * @returns {object} - all input plugins with path
+ * @returns {array} array of directory paths
  */
-exports.getPlugins = (globalConfig) => {
-  const existing = {};
+exports.getPluginsDirs = (globalConfig) => {
```
Comment (Snugug) [inline lib/utils.js]: This file appears to be an outlier amongst our work. Would like to see these functions as `const`s that then get `module.exports`-ed at the end

File: lib/utils.js [hunk 2]
```diff
@@ -55,34 +55,68 @@ exports.splitPop = (input, splitter) => {
...
   const root = exec('npm root').toString().trim();
-  let dir = _.get(globalConfig, 'content.plugins.directory', []);
+  const core = _.get(globalConfig, 'content.plugins.core.directory', []);
```
Comment (Snugug) [inline lib/utils.js]: * `content.plugins.directories.core` and `content.plugins.directories.local` (something like that). Unless we have other sets of configuration for `core` and `local` they should be under a shared `directory` object.
* What is `core`? What is `local`? Content types should really only care about additional find paths for plugins above the normal `npm` path.

File: lib/utils.js (hunk 1 above)
Comment (scottnath) [inline lib/utils.js]: fixed.

File: lib/utils.js (hunk 2 above)
Comment (scottnath) [inline lib/utils.js]: removed `core` and `local` as separate configs and plugin sources!

next issue please!

**Comment by Snugug** [comment]
This PR as it stands doesn't resolve the underlying issue in #96, which is being able to have 2 tiers of config both touch the `content.plugins.directory` config. There is a way to set up [sub-module configuration](https://github.com/lorenwest/node-config/wiki/Sub-Module-Configuration) which would resolve the particular issue for Punchcard, but would make Punchcard potentially brittle.

Options:

* Use Sub Module Configuration for the `config` module, and treat this PR as a just cleanup PR for loading in directories
* Use this PR (or close this/open a new one) that changes how Content Types gets configured, adding in a local config option on top of the `config` option, or removing the global alltogether

**Comment by scottnath** [comment]
@Snugug made changes to https://github.com/punchcard-cms/content-types/issues/96 to match this PR. Basically a cleanup/functional change for content types.

File: README.md [hunk 3]
```diff
@@ -168,3 +168,29 @@ content.only('my-awesome-content-type').then(type => {
...
 form.script // Validation and UX scripts, wrapped for browser via browserify
 ```
+
+### Input Plugins
+
+Forms are created using [input plugins](https://www.npmjs.com/search?q=input-plugin-) which are available as npm modules. Content-types automatically searches for these input plugins inside your application's `node_modules` directory, found using `npm root`.
+
+You can also use your own, non-npm-based input plugins in your application. Content-types uses [Node Config](https://github.com/lorenwest/node-config) for configuration, ([at least for now it does](https://github.com/punchcard-cms/content-types/issues/97)) so to use your own plugins you have to update the configuration for your app.
+
+You may use a string, or an array of strings. They must be full paths, not relative paths, or your application will error.
```
Comment (Snugug) [inline README.md]: An error will be thrown

**Comment by Snugug** [comment]
FWIW stories shouldn't be updated to match a PR

**Comment by Snugug** [comment]
Waiting on #96 being rewritten in a way that is user-centric

**Comment by scottnath** [comment]
#96 was re-written. This PR may not be needed. After https://github.com/punchcard-cms/content-types/pull/99 is merged, we may be able to just close this.

**Comment by Snugug** [comment]
Is this now deprecated?

**Comment by scottnath** [comment]
@Snugug yeah...lets let this go

**Pull request closed without merging**